test/TestApps/PushNotificationsDemoApp/main.cpp

```diff
@@ -0,0 +1,130 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+#include "pch.h"
+#include <wil/win32_helpers.h>
+#include <iostream>
+#include <time.h>
+#include <winrt/Windows.ApplicationModel.Background.h> // we need this for BackgroundTask APIs
+
+#define SIZE 26
+
+using namespace winrt::Microsoft::Windows::AppLifecycle;
+using namespace winrt::Microsoft::Windows::PushNotifications;
+using namespace winrt;
+using namespace winrt::Windows::Storage;
+using namespace winrt::Windows::Storage::Streams;
+using namespace winrt::Windows::Foundation;
+using namespace winrt::Windows::ApplicationModel::Activation;
+using namespace winrt::Windows::ApplicationModel::Background; // BackgroundTask APIs
+
+time_t ltime;
+char buf[SIZE];
+
+void RequestChannel()
+{
+    // Register the AAD RemoteIdentifier for the App to receive Push
+    auto channelOperation = PushNotificationManager::CreateChannelAsync(
+        winrt::guid("ccd2ae3f-764f-4ae3-be45-9804761b28b2"));
+
+    // Setup the inprogress event handler
+    channelOperation.Progress(
+        [](
+            IAsyncOperationWithProgress<PushNotificationCreateChannelResult, PushNotificationCreateChannelStatus> const& sender,
+            PushNotificationCreateChannelStatus const& args)
+        {
+            if (args.status == PushNotificationChannelStatus::InProgress)
+            {
+                // This is basically a noop since it isn't really an error state
+                std::cout << "Channel request is in progress." << std::endl << std::endl;
+            }
+            else if (args.status == PushNotificationChannelStatus::InProgressRetry)
+            {
+                LOG_HR_MSG(
+                    args.extendedError,
+                    "The channel request is in back-off retry mode because of a retryable error! Expect delays in acquiring it. RetryCount = %d",
+                    args.retryCount);
+            }
+        });
+
+    winrt::event_token pushToken;
+
+    // Setup the completed event handler
+    channelOperation.Completed(
+        [&](
```

Comment:
The outer method returns and destructs pushToken, but this `Completed` handler has a reference to that destroyed thing. Should the token go inside the completed handler below? Or be returned to the caller somehow to clean up? Or just don't bother capturing the event token returned at all?

(Note: capturing things by reference into lambdas appears to be a common hazard in this pull request; please double-check *all* instances of `[&](...)` and verify that the lifecycle of all things captured by-reference is correct.)

Reply:
In addition to lifetime problems both with the lambda and the created channel, this code is very non-idiomatic. Should just be a coroutine.

```cpp
winrt::Windows::Foundation::IAsyncOperation<PushNotificationChannel> RequestChannelAsync()
{
    // Register the AAD RemoteIdentifier for the App to receive Push
    auto channelOperation = PushNotificationManager::CreateChannelAsync(
        winrt::guid("ccd2ae3f-764f-4ae3-be45-9804761b28b2"));

    // Setup the inprogress event handler
    channelOperation.Progress(
        [](auto&& sender, auto&&args)
        {
            if (args.Status == PushNotificationChannelStatus::InProgress)
            {
                // This is basically a noop since it isn't really an error state
                std::cout << "Channel request is in progress." << std::endl << std::endl;
            }
            else if (args.Status == PushNotificationChannelStatus::InProgressRetry)
            {
                LOG_HR_MSG(
                    args.ExtendedError,
                    "The channel request is in back-off retry mode because of a retryable error! Expect delays in acquiring it. RetryCount = %d",
                    args.RetryCount);
            }
        });

    auto result = co_await channelOperation;

    if (result.Status() == PushNotificationChannelStatus::CompletedSuccess)
    {
        auto channelUri = result.Channel().Uri();

        std::cout << "channelUri: " << winrt::to_string(channelUri.ToString()) << std::endl << std::endl;

        auto channelExpiry = result.Channel().ExpirationTime();

        // Register Push Event for Foreground
        result.Channel().PushReceived([](const auto&, PushNotificationReceivedEventArgs const& args)
        {
            auto payload = args.Payload();

            // Do stuff to process the raw payload
            std::string payloadString(payload.begin(), payload.end());
            std::cout << "Push notification content received from FOREGROUND: " << payloadString << std::endl << std::endl;
            args.Handled(true);
        });
        // Caller's responsibility to keep the channel alive
        co_return result.Channel();
    }
    else if (result.Status() == PushNotificationChannelStatus::CompletedFailure)
    {
        LOG_HR_MSG(result.ExtendedError(), "We hit a critical non-retryable error with channel request!");
        co_return nullptr;
    }
    else
    {
        LOG_HR_MSG(result.ExtendedError(), "Some other failure occurred.");
        co_return nullptr;
    }
}
```